### PR TITLE
Trying to fix Substitution for core module propal & fcicheinter

### PR DESCRIPTION
**Pull request by zcattacz**
The module_parts['substitution'] never got set; this caused the customized substitution content never get loaded. Adding a work around here.
This PR doesn't seem to be the originally desired fix. 
Could someone help look into this ?

### CODE REVIEW

**Comment by zcattacz** [comment]
Heck. What a diff. Must forget to sync first :-( . Closing. Sorry for the noise.

**Comment by zcattacz** [comment]
Turns out that my code editing in web UI brought in a spacing mess (tabs/spaces).
the code in question is in 8dbc88f.

**Comment by eldy** [comment]
Sorry, i don't understand fix. The dir htfocs/core/substititions does not exists anymore in 3.4 and dev branch.

"substitions" is a system to allow a module to declare existence of a file  htdocs/mymodule/core/substitutions/function_mymodule.lib.php
but it is not used by core code (core code calls complete_substitutions_array function to complete an array by calling your module function into file htdocs/mymodule/core/substitutions/function_mymodule.lib.php but that's all).

**Comment by zcattacz** [comment]
Thanks for quick response. 
Yes I noticed that the dir doesn't exist. Actually the path is from the example in the Wiki ( http://wiki.dolibarr.org/index.php/Create_an_ODT_document_template )

```
Create a file into one of those locations: 

htdocs/core/substitutions/functions_mymodule.lib.php or htdocs/mymodule/core/substitutions/functions_mymodule.lib.php 
```

I created them in both location, but none of them got called. It seems that complete_substitutions_array function(https://github.com/zcattacz/dolibarr/blob/8dbc88f2a75acd588b04ec6854dce75c13e6b6a1/htdocs/core/lib/functions.lib.php#L3663-3667) needs modules_parts['substitutions'] to determine the path to the location of file "functions_mymodule.lib.php".

For core modules, modules_parts['substitutions'] is always empty, starting from the code block where 8dbc88f works on. In this case, complete_substitutions_array() always searches an empty path ""; never loads anything. 
I don't know how to safely determine a core module, so I just assign the the path to the module whose name contains "invoice" or "proposal".

Hope this help clear things up.

**Comment by eldy** [comment]
Hum, ok, i understand now why you suggest this.
The problem was not into the way you understood the tutorial, but was into the tutorial. Tutorial was correct for version <= 3.2 but wrong with 3.3+. I updated the wiki document. From 3.3 and 3.4, adding files into dolibarr directories is no more allowed, so to have a subsitution file working only the following path is allowed htdocs/mymodule/core/substitutions/functions_mymodule.lib.php.
Thats why the fix is not required because we don't want anymore to have substitution files into htdocs/core/substitutions/functions_mymodule.lib.php
I am really sorry to have see you spend time to fix this bug that in fact should not, but it was not competely a waste of time, because you point me that documentation was wrong and not up to date, and now fixed.

**Comment by zcattacz** [comment]
OK. but if I remembered correctly. that path doesn't work either. because 
complete_substitutions_array() use modules_parts['substitutions']  to find substitution script path.
and for core module that path is empty as "". so complete_substitutions_array() loads no substitution.

I might be wrong, let me try again with that path later.

No sorry pls. dolibarr worked very well for us in many aspects. Currently it suits our needs more than any other project I have tried out. Thank you for this great project.

**Comment by lrq3000** [comment]
Hello zcattacz,

I have quite contributed to the ODT substitution functionality and keep using it in my CustomFields module, thus I can maybe help you a bit.

The substitution works correctly but is sometimes a bit tricky. If you want a concrete example, you can check out the substitution file used by the opensource version of CustomFields here:
https://github.com/lrq3000/dolibarr_customfields/blob/3.3/htdocs/customfields/core/substitutions/functions_customfields.lib.php

Tell me if you have any other trouble using these substitutions.

PS: Ah and don't forget to declare substitutions activation in your module descriptor inside the $this->module_parts variable, like this:
https://github.com/lrq3000/dolibarr_customfields/blob/3.3/htdocs/customfields/core/modules/modCustomFields.class.php#L68

And then don't forget to disable/renable your module in the Dolibarr's module administration page, because you need to refresh your module descriptor in order for Dolibarr to take into account your new $this->module_parts !

**Comment by eldy** [comment]
You should not have need for substitution by core since there is no more core/substitutions/xxx files.
Let us know if you succeed or fails with the way after update of wiki (that should also be the way used by examples provided by lrq3000 I think).

**Comment by lrq3000** [comment]
Yes indeed the substitution files of CustomFields are not placed in the core but in its own module folder, as are all the other files used by CustomFields since the aim is to be totally independent of Dolibarr core code. Thus you can also use CustomFields as a good basis for your own module, and I've commented pretty much anything to provide good guidelines.

**Comment by zcattacz** [comment]
Hi folks, Thank you for the instructive replies.

I understand what you are suggesting now. First I need to create my own module. And by "mymodule" the wiki means PlaceYourModuleNameHere . Hoory now it works.

But I have some other problem that I need to fix.The VAT/TTC/HT items are just as annoying to us as how important they are to you. They are everywhere but not useful to us. And the same goes to ProfID.
Previously I simply comment them out in the source code. Now I am thinking to tag all the <tr> and <td> that the VAT is in with 'class="noVAT"' className and set a display:none for this class, in the theme.
What do you think? Any better alternatives? I haven't see any way to do this within my own module yet.
Is this kind of tagging useful to the project? There is a Disable VAT option in setup page, which don't hide vat items when turned on.

**Comment by eldy** [comment]
To answer correctly i need fist more information on your cae.
Can you provide more information on page you are annoying by VAT even after setting "Disable VAT" and provide a description on you activity and why VAT is annoying you ?
